lightning-invoice/src/utils.rs

```diff
@@ -21,6 +21,98 @@ use core::convert::TryInto;
 use core::ops::Deref;
 use core::time::Duration;
 
+#[cfg(feature = "std")]
+/// Utility to create an invoice that can be paid to one of multiple nodes, or a "phantom invoice."
+/// This works because the invoice officially pays to a fake node (the "phantom"). Useful for
+/// load-balancing payments between multiple nodes.
+///
+/// `channels` parameter:
+/// * Contains a list of tuples of `(phantom_node_scid, real_node_pubkey,
+///   vec[real_channel_details])`, where `phantom_node_scid` is unique to each entry
+/// * `phantom_node_scid`s are retrieved from [`ChannelManager::get_phantom_scid`]
+/// * Each `vec[real_channel_details]` is retrieved from [`ChannelManager::list_usable_channels`]
+/// * Contains channel info for all nodes participating in the phantom invoice
+/// * It is fine to cache `channels` and reuse it across invoices, as long as the data is
+///   updated when a channel becomes disabled or closes
+/// * Note that including too many channels here could result in making the invoice too long for QR
+///   code scanning
+///
+/// `payment_hash` and `payment_secret` come from [`ChannelManager::create_inbound_payment`] or
+/// [`ChannelManager::create_inbound_payment_for_hash`]. These values can be retrieved from any
+/// participating node.
+///
+/// See [`KeysInterface::get_phantom_secret`] for more requirements on supporting phantom node
+/// payments.
+///
+/// [`KeysInterface::get_phantom_secret`]: lightning::chain::keysinterface::KeysInterface::get_phantom_secret
+pub fn create_phantom_invoice<Signer: Sign, K: Deref>(
```

Comment:
How does it interferes with MPP ? Am I correct that you _can't_ MPP shard to each one of the participating nodes as they won't communicate among themselves shard reception ?

Maybe we could do that in the future, if we move the MPP accounting logic from `process_pending_htlc_forwards` behind an abstract interface.

Reply:
> How does it interferes with MPP ? Am I correct that you can't MPP shard to each one of the participating nodes as they won't communicate among themselves shard reception ?

Yep!

> Maybe we could do that in the future, if we move the MPP accounting logic from process_pending_htlc_forwards behind an abstract interface.

I suggested this to @TheBlueMatt but it seems to be too much of a compromise on safety guarantees, e.g. it could allow the user to cause us to release the preimage too early. Also a bit of a layer violation. 

Adding a comment clarifying this!

Reply:
> I suggested this to @TheBlueMatt but it seems to be too much of a compromise on safety guarantees, e.g. it could allow the user to cause us to release the preimage too early. Also a bit of a layer violation.

I think that's okay on the safety guarantees, if you have some two-phases commit protocol (you announce the HTLC reception to the coordinating server and then when all are received you release the preimage). IIRC, we might need such protocol if we want to coordinate well the channel monitor replicas. If it does happen, we can re-think MPP phantom invoice in the future.

Reply:
I mean its definitely possible to provide a complicated two-phase API for this, but I'm not sure that means we should. It'd definitely be a complicated API, and if a user wants to receive a large payment they can also just generate non-phantom invoices for large payments. We can revisit this if someone really needs it, though, indeed.